client/js/renderPreview.js

```diff
@@ -13,6 +13,18 @@ function renderPreview(preview, msg) {
 	}
 
 	preview.link = Handlebars.Utils.escapeExpression(preview.link);
+
+	const previewContainer = msg.find(`.preview[data-url="${preview.link}"]`);
+
+	// This is to fix a very rare case of rendering a preview twice
+	// This happens when a very large amount of messages is being sent to the client
+	// and they get queued, so the `preview` object on the server has time to load before
+	// it actually gets sent to the server, which makes the loaded preview sent twice,
+	// once in `msg` and another in `msg:preview`
+	if (previewContainer.data("loaded")) {
+		return;
+	}
```

Comment:
I think it would be nicer to check that `previewContainer.hasClass("toggle-type-loading")`, which is the meaning `loading` type was for in #1353, instead of adding a new attribute for this.